design-proposals/sriov-interfaces-mapping.md

```diff
@@ -0,0 +1,694 @@
+
+The outcome is a non-consistent association between the underlying VFs and the VM networks,
+leaving the VM in a state where the VFs are plugged into the domain wrongly.
+<br/>
+For example, when a SR-IOV interface is defined with a custom MAC address (e.g: 02:00:00:00:00), a custom guest PCI address of (e.g: 0000:01:01.1) and pointing to a NetworkAttachmentDefinition with VLAN configuration (e.g: 100).
+It is not guaranteed that the correct VF (with VLAN 100) will be plugged into the domain on the specified PCI address (0000:01:01.1).
+
+This issue also manifests when the VM interface boot order is set, the VM may end up booting from the wrong interface.
+
+A side effect of this issue is the VM status that may confuse its consumers, with its false advertisement, make them act incorrectly, and introduce other issues.
+<br/>
+For example, the cluster admin may mistakenly think that other network appliances are not configured correctly (e.g: router).
+Kubevirt should plug the correct VF with the correct properties into the domain,
+and the user inside the guest should be able to predict and assume a consistent SR-IOV NIC setup.
+
+## Goal
+Each VM SR-IOV interface should be associated with the correct VF.
+
+## Non-Goals
+Fix/change third-party components (e.g: Multus, SR-IOV CNI, etc..)
+
+## Existing Implementation Flow
+In order to discover and allocate VFs to Pods the [sriov-network-device-plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin#sr-iov-network-device-plugin) (from now on sriov-dp) and the [sriov-cni](https://github.com/k8snetworkplumbingwg/sriov-cni) are being used.
+<br/>
+The sriov-dp provides an [API to create resource pools](https://github.com/openshid/sriov-network-device-plugin#configurations) of VFs according to various properties,
+Each resource pool is labeled with a resource name.
+
+In the VMI spec, secondary networks point to the desired SR-IOV NetworkAttachmentDefinition object.
+The NetworkAttachmentDefinition is set with the Multus networks-status annotation:
+<br/>`k8s.v1.cni.cncf.io/resourceName=<resource name>`<br/>
+which later on passed down by Multus to the sriov-cni [[1]](https://github.com/k8snetworkplumbingwg/multus-cni/blob/a28f5cb56c79a582f5ea2b35a61b38f34b937930/examples/README.md#passing-down-device-information).
+```yaml
+kind: VirtualMachine
+  ...
+spec:
+  domain:
+    devices:
+      interfaces:
+      - name: sriovnet1
+        sriov: {}
+    ...
+  networks:
+  - name: sriovnet1
+    multus:
+      networkName: default/sriov-network-vlan100
+  ...
+```
+```yaml
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: sriov-network-vlan100
+  namespace: default
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: kubevirt.io/sriov_net
+spec:
+  config: |-
+    {
+        "type":"sriov",
+  ...
+```
+
+As part of Pod creation flow, kubelet executes the required device plugins according to the Pod spec, in our case the sriov-dp.
```

Comment:
Please consider giving a little background on how a Pod with SR-IOV is provisioned for the reader to better understand the problem:
1. kubelet calls the DP in order to allocate one or more VFs from a certain resource name.
2. Kubelet calls the container runtime in order to create the pod sandbox.
3. Container runtime calls Multus which does the association between network name, interface name, resource name and allocated PCI address.
4. Multus calls the SR-IOV CNI which configures the VF.
5. Multus adds the operation results in the Pod's `network-status` annotation.

The virt-launcher Pod has:
Several environment variables to map from:
1. Resource name to a list of allocated PCI addresses.
2. Network name to resource name.

Which are not enough to correctly associate VM's network and network interface with the actual VF that was previously configured and is represented as a PCI address.

Reply:
Thanks for the information you summarized it well.
Initially, I added this background after a few iterations of reviews, we came to the conclusion that it's too much information for the reader to grasp.
I agree it will give more background and thorough understanding, but note that for the common reader it may add more noise and distraction from the gist of the problem we try to fix.

Reply:
The main point of this explanation is that KubeVirt depends on an association that is made by Multus, and is reflected by it.
AFAIK, there is no other place to get this information from, other than the `network-status` annotation.

Reply:
I think the point that Kubevirt depends on Multus network-status annotation is expressed in the doc, isn't it?

Regarding the addendum, as you may already know, the flow of pod provisioning with secondary networks hides lots of nuances and various concepts such as but not only: device manager and device plugin API, CNI device info spec, and how Multus works.
Sure we can add a brief summary, but I am afraid it may leave the readers with a vague understanding (or even more confusion) since there are quite a lot of subjects to cover, and just mentioning it is not enough IMO.

As a side note, this is a proposal to fix certain functionality, it expects the reader to have a basic understanding of how Kubevirt and Multus integrate.
I tried to keep it detailed yet not overwhelming, compared to other proposals I think this proposal is pretty detailed.